### ISSUE TITLE
RadTube test failed with "Erroneous arithmetic operation"; resolved if set reconstructionOrder_ to 3

### ISSUE THREAD

**Issue by chongchonghe** [quokka-astro/quokka#679]
**Describe the bug**

RadTube on debug mode failed with "Erroneous arithmetic operation". This error can be traced back to PR #611 at the earliest. Is this expected? According to that PR, this error might have happened because NANs are encountered, but how do I debug this? I can't find any useful message in the error message. 

**To Reproduce**
1. In development branch, run RadTube test in Debug mode on Gadi. 
2. The program immediately fail with "Erroneous arithmetic operation". 

**Related**

This may have caused some tests to failure in PR #678

**Comment by BenWibking** [comment]
You can run it under gdb. Then it should show which line causes the problem.

**Comment by chongchonghe** [comment]
That's weird. It runs normally under gdb, but failed after Coarse STEP 1 when directly running the executable in Debug mode. 

`cmake ../.. -DCMAKE_BUILD_TYPE=Debug -DAMReX_SPACEDIM=1 -G Ninja`, then build and run RadTube:

```
----------------------------------------------------------
Initializing AMReX (24.06-26-g7418556773c7)...
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (24.06-26-g7418556773c7) initialized

[Warning] [Performance] The grid blocking factor (4) is too small for reasonable performance. It should be 32 (or greater) when running on GPUs, and 16 (or greater) when running on CPUs.

[Warning] [Performance] The maximum grid size (32) is too small for reasonable performance. It should be 128 (or greater) when running on GPUs, and 64 (or greater) when running on CPUs.

Coarse STEP 1 at t = 0 (0%) starts ...
Erroneous arithmetic operation
See Backtrace.0 file for details
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 8.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```

Using the same executable and run it with gdb:

```
Initializing AMReX (24.06-26-g7418556773c7)...
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (24.06-26-g7418556773c7) initialized

[Warning] [Performance] The grid blocking factor (4) is too small for reasonable performance. It should be 32 (or greater) when running on GPUs, and 16 (or greater) when running on CPUs.

[Warning] [Performance] The maximum grid size (32) is too small for reasonable performance. It should be 128 (or greater) when running on GPUs, and 64 (or greater) when running on CPUs.

Coarse STEP 1 at t = 0 (0%) starts ...

Coarse STEP 2 at t = 2.481665465e-09 (0.078125%) starts ...

Coarse STEP 3 at t = 4.954290841e-09 (0.1559654101%) starts ...

Coarse STEP 4 at t = 7.426885242e-09 (0.2338048451%) starts ...

Coarse STEP 5 at t = 9.899439587e-09 (0.3116430191%) starts ...

Coarse STEP 6 at t = 1.237194844e-08 (0.3894797608%) starts ...

Coarse STEP 7 at t = 1.484440862e-08 (0.4673149704%) starts ...
```

**Comment by BenWibking** [comment]
Ah...a heisenbug :) 

Does it produce any Backtrace.* files?

**Comment by chongchonghe** [comment]
Yes. Here are the backtrace files on macOS and on Gadi:

macOS:

```
 0: amrex::BLBackTrace::print_backtrace_info(__sFILE*) (in test_radiation_tube) (AMReX_BLBackTrace.cpp:190)

 1: amrex::BLBackTrace::handler(int) (in test_radiation_tube) (AMReX_BLBackTrace.cpp:98)

 2: _sigtramp (in libsystem_platform.dylib) + 56

 3: void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int)::'lambda'(int, int, int, int, int)::operator()(int, int, int, int, int) const (in test_radiation_tube) (hydro_system.hpp:676)

 4: std::__1::enable_if<IsFabArray<amrex::MultiFab>::value, void>::type amrex::experimental::detail::ParallelFor<amrex::MultiFab, void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int)::'lambda'(int, int, int, int, int)>(amrex::MultiFab const&, amrex::IntVectND<1> const&, int, amrex::IntVectND<1> const&, bool, void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int)::'lambda'(int, int, int, int, int) const&) (in test_radiation_tube) (AMReX_MFParallelForC.H:51)

 5: std::__1::enable_if<IsFabArray<amrex::MultiFab>::value, void>::type amrex::experimental::ParallelFor<amrex::MultiFab, void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int)::'lambda'(int, int, int, int, int)>(amrex::MultiFab const&, amrex::IntVectND<1> const&, int, void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int)::'lambda'(int, int, int, int, int)&&) (in test_radiation_tube) (AMReX_MFParallelFor.H:150)

 6: void HydroSystem<TubeProblem>::FlattenShocks<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, int, int) (in test_radiation_tube) (hydro_system.hpp:689)

 7: void RadhydroSimulation<TubeProblem>::hydroFluxFunction<(FluxDir)0>(amrex::MultiFab const&, amrex::MultiFab&, amrex::MultiFab&, amrex::MultiFab&, amrex::MultiFab&, amrex::MultiFab const&, amrex::MultiFab const&, amrex::MultiFab const&, int, int) (in test_radiation_tube) (RadhydroSimulation.hpp:1504)

 8: RadhydroSimulation<TubeProblem>::computeHydroFluxes(amrex::MultiFab const&, int, int) (in test_radiation_tube) (RadhydroSimulation.hpp:1438)

 9: RadhydroSimulation<TubeProblem>::advanceHydroAtLevel(amrex::MultiFab&, amrex::YAFluxRegisterT<amrex::MultiFab>*, amrex::YAFluxRegisterT<amrex::MultiFab>*, int, double, double) (in test_radiation_tube) (RadhydroSimulation.hpp:1098)

10: RadhydroSimulation<TubeProblem>::advanceHydroAtLevelWithRetries(int, double, double, amrex::YAFluxRegisterT<amrex::MultiFab>*, amrex::YAFluxRegisterT<amrex::MultiFab>*) (in test_radiation_tube) (RadhydroSimulation.hpp:945)

11: RadhydroSimulation<TubeProblem>::advanceSingleTimestepAtLevel(int, double, double, int) (in test_radiation_tube) (RadhydroSimulation.hpp:676)

12: AMRSimulation<TubeProblem>::timeStepWithSubcycling(int, double, int) (in test_radiation_tube) (simulation.hpp:1282)

13: AMRSimulation<TubeProblem>::evolve() (in test_radiation_tube) (simulation.hpp:876)

14: problem_main() (in test_radiation_tube) (test_radiation_tube.cpp:292)

15: main (in test_radiation_tube) (main.cpp:54)

16: start (in dyld) + 2360


===== TinyProfilers ======
AMRSimulation::evolve()
AMRSimulation::timeStepWithSubcycling()
RadhydroSimulation::advanceSingleTimestepAtLevel()
REG::HydroSolver
RadhydroSimulation::advanceHydroAtLevel()
RadhydroSimulation::computeHydroFluxes()
```

On Gadi:

```
Host Name: gadi-login-08.gadi.nci.org.au
=== If no file names and line numbers are shown below, one can run
            addr2line -Cpfie my_exefile my_line_address
    to convert `my_line_address` (e.g., 0x4a6b) into file name and line number.
    Or one can use amrex/Tools/Backtrace/parse_bt.py.

=== Please note that the line number reported by addr2line may not be accurate.
    One can use
            readelf -wl my_exefile | grep my_line_address'
    to find out the offset for that line.

 0: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x6243a6]
    amrex::BLBackTrace::print_backtrace_info(_IO_FILE*) at ??:?

 1: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x626580]
    amrex::BLBackTrace::handler(int) at ??:?

 2: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x722843]
    amrex::Gpu::Device::Initialize() at ??:?

 3: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x59342e]
    amrex::Initialize(int&, char**&, bool, ompi_communicator_t*, std::function<void ()> const&, std::ostream&, std::ostream&, void (*)(char const*)) at ??:?

 4: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x42fb76]
    main at ??:?

 5: /lib64/libc.so.6(__libc_start_main+0xe5) [0x7fb7df427d85]

 6: /home/120/ch1506/softwares/quokka/quokka-mac/build/build-gpu-3D/src/RadTube/test_radiation_tube() [0x4366ce]
    _start at ??:?


===== TinyProfilers ======
```

**Comment by BenWibking** [comment]
What compiler are you using?

**Comment by BenWibking** [comment]
I don't have any good ideas of how to debug this. Maybe ask on the AMReX slack?

**Comment by chongchonghe** [comment]
> What compiler are you using?

On macOS i'm using `Apple clang version 15.0.0` and on gadi I'm using `module load pbs gcc/11.1.0 openmpi/4.1.4 intel-mkl/2022.2.0 python3/3.11.0 hdf5/1.10.7`

**Comment by chongchonghe** [comment]
> I don't have any good ideas of how to debug this. Maybe ask on the AMReX slack?

This error only occurred after #611 , so any idea what could be going wrong?

**Comment by BenWibking** [comment]
> > I don't have any good ideas of how to debug this. Maybe ask on the AMReX slack?
> 
> This error only occurred after #611 , so any idea what could be going wrong?

That PR only enables floating-point exception trapping. So it's possible there was an NAN value being produced somewhere that was being missed before, or there could be a compiler bug that is causing spurious FPEs. It is a bit strange that this does not show up in CI.

**Comment by BenWibking** [comment]
Did you add these compiler flags: https://github.com/quokka-astro/quokka/pull/611#issuecomment-2203814570?

**Comment by BenWibking** [comment]
This might be the same issue: https://github.com/quokka-astro/quokka/pull/611#issuecomment-2206986495.

I thought it was an Intel compiler issue, since I only saw it there, and couldn't reproduce with any other compilers.

**Comment by chongchonghe** [comment]
> This might be the same issue: [#611 (comment)](https://github.com/quokka-astro/quokka/pull/611#issuecomment-2206986495).
> 
> I thought it was an Intel compiler issue, since I only saw it there, and couldn't reproduce with any other compilers.

Yes, I only see it in intel compiler on gadi and I don't see it in Apple clang.

**Comment by chongchonghe** [comment]
> Did you add these compiler flags: [#611 (comment)](https://github.com/quokka-astro/quokka/pull/611#issuecomment-2203814570)?

How do I add `-ffp-exception-behavior=maytrap` compiler flag exactly?

**Comment by BenWibking** [comment]
https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_FLAGS.html#variable:CMAKE_%3CLANG%3E_FLAGS

**Comment by chongchonghe** [comment]
So, we need to add `-ffp-exception-behavior=maytrap` to cmake-macos.yml. I just did that in #680 . I'll see what happens.

**Comment by chongchonghe** [comment]
Resolved in #678 . Redirect the real problem to #681
